flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandler.java

```diff
@@ -179,8 +178,14 @@ class ResultSubpartitionRecoveredStateHandler
     public BufferWithContext<BufferBuilder> getBuffer(ResultSubpartitionInfo subpartitionInfo)
             throws IOException, InterruptedException {
         // request the buffer from any mapped subpartition as they all will receive the same buffer
-        final List<CheckpointedResultSubpartition> channels = getMappedChannels(subpartitionInfo);
-        BufferBuilder bufferBuilder = channels.get(0).requestBufferBuilderBlocking();
+        ResultPartitionWriter writer = writers[subpartitionInfo.getPartitionIdx()];
+        if (!(writer instanceof CheckpointedResultPartition)) {
```

Comment:
No, I don't have any good suggestions for now. I just wanted to emphasize that `ResultSubpartitionRecoveredStateHandler` uses `ResultPartitionWriter` for the writer while it actually needs `CheckpointedResultPartition`. As a result, it requires the explicit cast to this type. 
So it doesn't look clean solution but since it was here before and I don't actually think that we should try to improve it in this fix. We just can leave it as is.